[Diagram] Member 이벤트 다이어그램 작성

vwx12345 opened the pull request:
### 💻 작업 내용
- 자기 정보 조회 다이어그램 추가
- 닉네임으로 해당 유저 존재 체크 다이어그램 추가
- 프로필 업데이트 다이어그램 추가
- 회원 탈퇴 다이어그램 추가

### ✨ 리뷰 포인트
- getMyProfile 에서 domain과 관련된 mapper를 작성했는데 이 부분도 다이어그램에 반영해야할지 고민입니다.

### 📝 메모
- 조회는 따로 DomainEventPublisher에 publish 생성하지 않았습니다.

### 🎯 관련 이슈
#5

Commit:
docs(diagram): getMyProfile.puml 수정

Commit:
docs(diagram): emailSignUp.puml 삭제
- emailSignUp.puml의 경우, member context가 아님으로 삭제

Commit:
docs(diagram): oauthSignUp.puml 삭제
-  oauthSignUp.puml의 경우, member context가 아님으로 삭제

Commit:
docs(diagram): nicknameExistsCheck.puml 수정
- Service도 Query와 Command로 나눔에 따라 MemberService -> MemberQueryService로 변경

Commit:
docs(diagram): 회원 프로필 조회 시나리오 시퀀스 다이어그램 추가

Commit:
docs(diagram): 회원 정보 수정 시나리오 시퀀스 다이어그램 추가

Commit:
docs(diagram): 회원 탈퇴 시나리오 시퀀스 다이어그램 추가

Commit:
Merge branch 'develop' of https://github.com/SimpleBoardTeam/simpleBoard-server into document/diagram-event-member/#5

codingbaraGo commented on document/puml/event/memberContext/memberAggregate/getMyProfile.puml:
```diff
@@ -0,0 +1,24 @@
+MemberQueryService -> MemberQueryRepository : findById(memberId)
+MemberQueryRepository --> MemberQueryService : Member
+
+MemberQueryService -> Member : toProfileResult()
```
to 보다는 get이 어울리는 상황이 아닐까 하네요
to는 argument로 데이터를 전달하거나 argument의 데이터를 변환하는 느낌이에요

vwx12345 replied:
> to 보다는 get이 어울리는 상황이 아닐까 하네요 to는 argument로 데이터를 전달하거나 argument의 데이터를 변환하는 느낌이에요

아, 연결이 잘못되어 있습니다. 이 함수는 Member를 서비스 계층에서 사용할 ProfileResult로 매핑하기 위한 용도입니다.
다만 아직 ProfileResult를 사용할 일이 없어서, 오버엔지니어링 같다는 생각에 제거할지도 고민 중입니다.

codingbaraGo commented on document/puml/event/memberContext/memberAggregate/getMyProfile.puml:
```diff
@@ -0,0 +1,24 @@
+MemberQueryService -> Member : toProfileResult()
+Member --> MemberQueryService : MemberProfileResult
+
+MemberQueryService -> MemberProfileMapper : toResponse(MemberProfileResult)
```
MemberProfileMapper는 일종의 converter인가요?
그렇다면 converter가 필요한 백로그 하나당 하나의 converter 클래스를 생성하는 1대1 정책인가요?

vwx12345 replied:
> MemberProfileMapper는 일종의 converter인가요? 그렇다면 converter가 필요한 백로그 하나당 하나의 converter 클래스를 생성하는 1대1 정책인가요?

네, MemberProfileMapper는 일종의 converter 개념으로 작성했습니다.
처음엔 Member → ProfileResult → MemProfileResponse로 나누어 변환하려 했는데, 지금 상황에서는 굳이 단계별 분리를 할 필요가 없을 것 같아 단순화할지 고민 중입니다.

백로그 하나당 하나의 converter를 두는 1:1 정책까지는 아직 고려하지 않았고, 필요에 따라 유연하게 구성하려고 합니다.

codingbaraGo commented on document/puml/event/memberContext/memberAggregate/nicknameExistsCheck.puml:
```diff
@@ -0,0 +1,15 @@
+MemberQueryRepository --> MemberQueryService : true / false
+
+MemberQueryService --> MemberController : true / false
+MemberController --> User : HTTP 200 OK + JSON (exists: true / false)
```
응답으로 200 / 403은 어떤가요?
굳이 반환 DTO 정의할 필요도 없고, 프론트 입장에서도 json 파일을 역직렬화 할 필도 없이 코드만으로 플로우를 나눌 수 있지 않을까 하네요

vwx12345 replied:
> 응답으로 200 / 403은 어떤가요? 굳이 반환 DTO 정의할 필요도 없고, 프론트 입장에서도 json 파일을 역직렬화 할 필도 없이 코드만으로 플로우를 나눌 수 있지 않을까 하네요

존재하지 않는 게 꼭 에러는 아니라 판단해 true/false로 응답하도록 했습니다.
200/403으로 나누는 방식도 의미 있지만, 이 API는 단순 존재 여부 확인이라 Boolean 응답이 더 직관적이라 생각했습니다.

codingbaraGo commented on document/puml/event/memberContext/memberAggregate/updateMyProfile.puml:
```diff
@@ -0,0 +1,32 @@
+MemberCommandService -> MemberCommandRepository : findById(memberId)
+MemberCommandRepository --> MemberCommandService : Member
+
+alt newNickname != null
+    MemberCommandService -> Member : updateNickname(newNickname)
+    MemberCommandService -> DomainEventPublisher : publish(NicknameUpdatedEvent)
```
프로필 업데이트 이벤트에 닉네임 중복 체크 관련 처리가 있나요?
닉네임 중복 체크 이벤트에서 결과값으로 토큰같은걸 발급하지는 않는 듯 한데
현재 형태는 유저가 포스트맨으로 프로필 업데이트 리퀘스트를 쏘면 중복 닉네임이 가능해지거나, DB에 제약 조건을 걸어놨다면 DB에서 에러가 발생하지 않을까 하네요

vwx12345 replied:
> 프로필 업데이트 이벤트에 닉네임 중복 체크 관련 처리가 있나요? 닉네임 중복 체크 이벤트에서 결과값으로 토큰같은걸 발급하지는 않는 듯 한데 현재 형태는 유저가 포스트맨으로 프로필 업데이트 리퀘스트를 쏘면 중복 닉네임이 가능해지거나, DB에 제약 조건을 걸어놨다면 DB에서 에러가 발생하지 않을까 하네요

해당 검증은 Member의 updateNickname 함수에서 처리해야 한다고 판단해 이벤트에는 포함하지 않았습니다.
updateNickname을 여기서 처리하게 되면 updatePassword 같은 다른 업데이트 로직도 함께 다뤄야 해서, 다이어그램이 지나치게 복잡해질 수 있다고 생각해 어떻게 처리할지 고민입니다.

codingbaraGo commented on document/puml/event/memberContext/memberAggregate/updateMyProfile.puml:
```diff
@@ -0,0 +1,32 @@
+end
+
+alt newPassword != null
+    MemberCommandService -> Member : updatePassword(newPassword)
+    MemberCommandService -> DomainEventPublisher : publish(PasswordUpdatedEvent)
```
혹시 password 변경 전 old password 체크 로직은 고려해 보셨나요?
꼭 필요하진 않을 것 같긴 하네요

vwx12345 replied:
> 혹시 password 변경 전 old password 체크 로직은 고려해 보셨나요? 꼭 필요하진 않을 것 같긴 하네요

위에 내용과 중복되나 old password 뿐만 아니라 유효성도 검증 등 처리할 로직이 많을 것 같아 어떻게 처리할지 고민입니다.

codingbaraGo commented on document/puml/event/memberContext/memberAggregate/withdrawMember.puml:
```diff
@@ -0,0 +1,20 @@
+
+MemberCommandService -> MemberCommandRepository : findById(memberId)
+MemberCommandRepository --> MemberCommandService : Member
+MemberCommandService -> MemberCommandRepository : delete(member)
```
맴버는 하드 딜리트군요.
일반적으로 회원 탈퇴 후 일정 기간은 복구가 가능하도록 만들어 놓으니 이 부분은 m2나 3정도에 같이 고민해봐요

vwx12345 replied:
erd 기준으로 isdelete 와 같은 컬럼이 없어서 하드 딜리트로 생각했습니다. 추후에 같이 상의해보는 게 좋을 것 같아요!

codingbaraGo commented on document/puml/event/memberContext/memberAggregate/withdrawMember.puml:
```diff
@@ -0,0 +1,20 @@
+MemberCommandService -> DomainEventPublisher : publish(MemberDeletedEvent)
+
+MemberCommandService --> MemberController : 탈퇴 성공
+MemberController --> User : HTTP 204 No Content
```
회원 탈퇴시의 화면 플로우를 생각해 보면 

로그인 되어있는 체로 마이페이지 화면에 있었음
-> 회원 탈퇴 버튼 누름
-> 회원 탈퇴 & 로그아웃 됨
-> 메인페이지로 리다이랙트

이렇게 흐를테니 No Content를 받으면 프론트에서는 메인페이지로 리다이렉트 시켜줘야겠네요

vwx12345 replied:
미쳐 생각하지 못한 부분이네요! 그렇게 처리하는게 훨씬 좋을 것 같습니다. 반영해서 다시 pr 하겠습니다.

codingbaraGo reviewed (commented):
고생하셨습니다!
PR페이지 우측에 리뷰어를 등록해주시면 더욱 빠른 리뷰가 가능합니다

vwx12345 committed:
Merge remote-tracking branch 'origin/develop' into document/diagram-event-member/#5

vwx12345 committed:
"chore(directory): auth context domain, board event 다이어그램 전체 구조 변경으로 삭제"

vwx12345 committed:
"chore(directory): member Context 를 external API 하위로 위치 변경
- member Context를 board Context 안에 넣는 실수를 하여 수정

Commit:
docs(diagram): MemberProfileMapper 제거

Commit:
docs(diagram): 닉네임 변경 기능 추가
- NicknameValidator를 통한 유효성 검사 수행
- 닉네임 변경 시 NicknameUpdatedEvent 발행

Commit:
docs(diagram): 회원 탈퇴 로직 변경
- auth에서의 삭제는 동기로 처리해야한다고 판단해서 authController 에 Delete api를 보내는 방식으로 구현
- 삭제는 hard delete 보다는 soft delete 가 서비스적으로 나을 것이라 판단하여 soft delete로 구현
- member 가 삭제됨에 따라 부가적으로 처리해야할 작업이 있다고 판단(ex 탈퇴 회원 nickname 제거 등)하여 event 발생 처리

Commit:
docs(diagram): getMyProfile 타이틀 등록

Commit:
docs(diagram): getProfileBYId 구현

Commit:
docs(diagram): checkMemberExists 로 이름 변경

Commit:
Merge remote-tracking branch 'origin/develop' into document/diagram-event-member/#5

vwx12345 committed:
Merge branch 'develop' into document/diagram-event-member/#5

vwx12345 committed:
chore(diagram): auto-generate SVGs

vwx12345 committed:
docs(diagram): getProfileById 와 같은 구조로 변경

vwx12345 committed:
docs(diagram): member domain 까지만 가도록 변경

vwx12345 committed:
chore(diagram): 잘못된 표현 수정
- 인증 세션이 아닌 인증 엔티티인데 잘못 적혀있어 수정

codingbaraGo commented on document/puml/event/externalAPI/memberContext/memberAggregate/updateNickname.puml:
```diff
@@ -0,0 +1,32 @@
+activate Member
+MemberCommandRepository --> MemberCommandService : Member
+
+MemberCommandService -> Member : updateNickname(newNickname)
+
+alt 닉네임이 유효하고 중복 아님
+Member --> MemberCommandService : 완료
+MemberCommandService -> DomainEventPublisher : NicknameUpdatedEvent
+deactivate Member
+MemberCommandService -> MemberCommandRepository : save(Member)
+MemberCommandService --> MemberController : OK
+MemberController --> User : 200 OK
```
현재의 형태를 보면 Member 엔티티 내에서 닉네임 중복검사를 하는 구조로 보이는데 맞나요?
엔티티에서 레포지토리에 직접 접근하려면 Member 객체에 레포지토리의 레퍼런스도 있어야하고, Member 엔티티가 해당 인스턴스 내부의 정보 관리가 아닌, 외부의 다른 데이터들에 대한 관리를 담당하게 되는데
책임의 범위가 너무 커지는건 아닐까 하는 생각이 드네요

vwx12345 replied:
저도 책임이 너무 크다고 생각해서 member 내부에서 NicknamePolicy 라는 별도의 책임을 가진 인터페이스를 구현하여 해결하고자 했습니다. 하지면 현재 NicknamePolicy를 언제 주입해줄지에 대한 고민 중입니다. 결정되면 다시 pr 올리겠습니다.

codingbaraGo reviewed (approved):
고생하셨습니다!
그런데 비밀번호 변경은 빠졌나요?

vwx12345 committed:
Merge branch 'develop' into document/diagram-event-member/#5

vwx12345 commented:
> 고생하셨습니다! 그런데 비밀번호 변경은 빠졌나요?

비밀번호는 auth context에서 담당하도록 변경했습니다.

vwx12345 committed:
chore(diagram): auto-generate SVGs
